### PR TITLE
Request controller

### LINKED ISSUE #6: BE: Request Controllers
\- GET /requests: list of requests for logged in user
\- POST /request: Create a new request
\- UPDATE /request/:id : Update request with approved or decline

### LINKED ISSUE #7: BE: Request Model
\- user\_id (user who is requesting)
\- sitter\_id (pet sitter user id)
\- start (date requested to start)
\- end (date requested to end)
\- accepted (default False) 
\- declined (default False)
\- paid (default False)

### PR DESCRIPTION
**Pull request by g4rry420**
### What this PR does (required):
Created the controller and routes for Request model.
#6 

### Screenshots / Videos (required):
 N/A

### Any information needed to test this feature (required):
Currently, Request model is work-in-progress by Joel, we can test this after model is completed.

### Any issues with the current functionality (optional):
No

### CODE REVIEW

File: server/controllers/request.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        })
+      }
+
+      req.request = request;
+      next();
+
+    })
+}
+
+exports.getAllRequest = (req, res) => {
```
Comment (dproc96) [inline server/controllers/request.js:19]: I'm not sure I see a use case for this route, would it make more sense to get all requests for the logged in user

File: server/controllers/request.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        })
+      }
+
+      return res.status(200).json({ requests });
+
+    })
+}
+
+exports.createRequest = (req, res) => {
+  const request = new Request(req.body);
```
Comment (dproc96) [inline server/controllers/request.js:48]: This request hasn't been linked to the logged in user in any way, if the user id is sent in the body that means that User A could create a request from User B to User C which we do not want to be possible

File: server/controllers/request.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
+// require Request Model
+const Request = {};
+
+exports.getRequestById = (req, res, next, id) => {
```
Comment (dproc96) [inline server/controllers/request.js]: Should we add some validation to this controller to see if the logged in user is a part of this request?

File: server/controllers/request.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    return res.status(200).json({
+      request
+    })
+
+  })
+}
+
+exports.updateRequest = (req, res) => {
```
Comment (dproc96) [inline server/controllers/request.js:63]: This route should probably confirm that the logged in user is the one receiving the request

File: server/routes/request.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+const protect = require('../middleware/auth');
+const { 
+  getAllRequest, 
+  createRequest, 
+  updateRequest, 
+  getRequestById 
+} = require('../controllers/request');
+
+router.param('requestId', getRequestById);
```
Comment (dproc96) [inline server/routes/request.js:13]:
```suggestion
router.param('/:requestId', getRequestById);

```

**Review by dproc96 — COMMENTED** [review]
Had some comments below on some security concerns

File: server/controllers/request.js (hunk 3 above)
Comment (g4rry420) [reply server/controllers/request.js]: I am sorry, I didn't understand this one. How can we add validation for param ? Please let me know.

File: server/controllers/request.js (hunk 1 above)
Comment (g4rry420) [reply server/controllers/request.js:19]: okay, I will make changes to it.

File: server/controllers/request.js (hunk 4 above)
Comment (g4rry420) [reply server/controllers/request.js:63]: I am sorry, I didn't understand. I did confirm by adding the `protect` middleware that the logged in user is receiving the request. Can you please clarify this for me ? If you meant by pull right request from the database, I think specifying the `_id` will be sufficient or we can also fetch the request from database using the `user_id`. Please, let me know about this.

File: server/routes/request.js (hunk 5 above)
Comment (g4rry420) [reply server/routes/request.js:13]: I am also not sure about this suggestion. As per the express [docs](https://expressjs.com/en/4x/api.html#router.param), we should define params like this. Can you please clarify this me ?

File: server/controllers/request.js (hunk 2 above)
Comment (g4rry420) [reply server/controllers/request.js:48]: got it. I will add validation for the `createRequest`.

**Comment by g4rry420** [comment]
@dproc96 Thank you Dan for your thorough review. I have made some changes, please let me know about them

File: server/controllers/request.js (hunk 4 above)
Comment (g4rry420) [reply server/controllers/request.js:63]: Never mind, I get it.

File: server/controllers/request.js (hunk 3 above)
Comment (sina-jamshidi) [reply server/controllers/request.js]: I think the point was that if we fetch a request by Id, we should make sure it belongs to the user requesting it

**Review by sina-jamshidi — APPROVED** [review]
This is good to merge

**Pull request merged**